### PR TITLE
Hypergraphs building models

### PR DESCRIPTION
**Pull request by spagnuolocarmine**
Support for hypergraphs generation in the model.jl file.

This version provides the following models:
- **Random**, given two integer values (_n_ and _m_), generates a completely random hypergraph with _n_ nodes and _m_ hyperedges;
 - **Random k-uniform**,  given three integer values (_n_,_m_,_k_), generates a random hypergraph with _n_ nodes and _m_ hyperedges, where each hyperedge has a size equal to _k_;
 - **Random d-regular**,  given three integer values (_n_,_m_,_d_), generates a random hypergraph with _n_ nodes and _m_ hyperedges, where each node has degree equal to _d_;
 - **Random preferential**,  given an integer value (_n_) and a real probability value (_p_), generates a random hypergraph with _n_ nodes, where each node has degree proportional to a preferential attachment probability _p_. As described in the paper: Avin, C., Lotker, Z., and Peleg, D.Random preferential attachment hyper-graphs.Computer Science 23 (2015).

**_Usage_**

```julia
#generates a random Hypergraph{Bool}(5,5)
h = random_model(5,5) 

#generates a random k-uniform Hypergraph{Bool}(5,5)
h = random_kuniform_model(5, 5, 3)

#generates a random d-regular Hypergraph{Bool}(5,5)
h = random_dregular_model(5, 5, 3) 

#generates a random with preferential attachment 
#Hypergraph{Bool}(20,m), where m is the number of hyperedges computed by the algorithm
h = random_preferential_model(20, 0.5) 
```

### CODE REVIEW

File: src/models.jl [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+"""
+function random_model(nVertices::Int, nEdges::Int)
+    mx = Matrix{Union{Nothing,Bool}}(nothing, nVertices,nEdges)
+    for e in 1:size(mx,2)
+        nv = rand(1:size(mx,1))
+        mx[sample(1:size(mx,1), nv;replace=false), e] .= true
+    end
+
+    h = Hypergraph(mx)
+    if all(length.(h.v2he) .> 0)
```
Comment (bkamins) [inline src/models.jl:29]: why do we add this restriction?

File: src/models.jl [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+"""
+function random_kuniform_model(nVertices::Int, nEdges::Int, k::Int)
+    mx = Matrix{Union{Nothing,Bool}}(nothing, nVertices,nEdges)
+    for e in 1:size(mx,2)
+        nv = k
+        mx[sample(1:size(mx,1), nv;replace=false), e] .= true
+    end
+
+    h = Hypergraph(mx)
+    if all(length.(h.he2v) .== k)
```
Comment (bkamins) [inline src/models.jl:53]: here it should probably an assertion - right?

File: src/models.jl [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+function random_dregular_model(nVertices::Int, nEdges::Int, d::Int)
+    mx = Matrix{Union{Nothing,Bool}}(nothing, nVertices,nEdges)
+
+    for v in 1:size(mx,1)
+        ne = d
+        mx[v, sample(1:size(mx,2), ne;replace=false)] .= true
+    end
+
+    h = Hypergraph(mx)
+    if all(length.(h.v2he) .== d)
```
Comment (bkamins) [inline src/models.jl:80]: again - this probably should be an assertion.

File: src/models.jl [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+        * preferential-attachment
+"""
+
+
+"""
+    random_model(nVertices, nEdges)
+
+Generate a *random* hypergraph without any structural constraints.
+
+**The Algorithm**
```
Comment (bkamins) [inline src/models.jl:15]: maybe this will be better:
```suggestion
# The Algorithm
```
same below

File: src/models.jl (hunk 1 above)
Comment (spagnuolocarmine) [reply src/models.jl:29]: Here, we suppose that a node needs to be into at least one hyperedge.

File: src/models.jl (hunk 2 above)
Comment (spagnuolocarmine) [reply src/models.jl:53]: We use it as control, to generate again the random graph, not to assert an error

File: src/models.jl [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+The algorithm starts with a random graph with 5 nodes and 5 edges. For this reason,
+the generated random graph has at least 5 nodes and 5 edges.
+It iteratively adds a node or a edge, according to a given parameter *p*,
+which defines the probability of creating a new node or a new hyperedge.
+
+More in detail, the connections with the new node/hyperedge are generated according to
+a preferential attachment policy defined by _p_.
+"""
+function random_preferential_model(nVertices::Int, p::Real)
+    H₀ = random_model(5,5)
```
Comment (bkamins) [inline src/models.jl:104]: why is `H₀` needed?

File: src/models.jl (hunk 3 above)
Comment (spagnuolocarmine) [reply src/models.jl:80]: The same for the random model, but here we are looking for a random graph with each node has degree equal to d

File: src/models.jl [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+            push!(Y,v=>true)
+            for v in nextNodes(H,y-1)
+                push!(Y,v)
+            end
+        else
+            for v in nextNodes(H,y)
+                push!(Y,v)
+            end
+        end
+            SimpleHypergraphs.add_hyperedge!(H, vertices=Y)
```
Comment (bkamins) [inline src/models.jl:121]: indentation

File: src/models.jl (hunk 5 above)
Comment (spagnuolocarmine) [reply src/models.jl:104]: is needed in the algorithm that we used...we also marked the paper in the doc

File: src/models.jl (hunk 6 above)
Comment (spagnuolocarmine) [reply src/models.jl]: fixed it

File: src/models.jl [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+a preferential attachment policy defined by _p_.
+"""
+function random_preferential_model(nVertices::Int, p::Real)
+    H₀ = random_model(5,5)
+    H = H₀
+    while nhv(H) < nVertices
+        r = rand()
+        y = rand(1:nhv(H))
+        Y = Dict{Int,Bool}()
+        if r < p
```
Comment (bkamins) [inline src/models.jl:110]: so we either add a vertex or a hyperedge only?

File: test/runtests.jl [hunk 8]
```diff
@@ -244,6 +267,7 @@ end;
     @test minimum([sum((h_from_g .== true)[:,n]) for n in 1:6] .== 2)
     @test LightGraphs.modularity(g,[1,1,2,2,3,3,4,4]) ≈ modularity(h_from_g, Set.([[1,2],[3,4],[5,6],[7,8]]))
     @test LightGraphs.SimpleGraphs.fadj(g) == LightGraphs.SimpleGraphs.fadj(TwoSectionView(h_from_g))
+
```
Comment (bkamins) [inline test/runtests.jl:270]: this newline is probably not needed

File: src/models.jl (hunk 7 above)
Comment (spagnuolocarmine) [reply src/models.jl:114]: yes correct

File: src/models.jl (hunk 1 above)
Comment (bkamins) [reply src/models.jl:30]: I understand, but why is this restriction added?

File: src/models.jl (hunk 2 above)
Comment (bkamins) [reply src/models.jl:55]: in what case it is possible that the edge does not have `k` vertices in this algorithm?

File: src/models.jl (hunk 3 above)
Comment (bkamins) [reply src/models.jl:83]: again - is this possible that this check fails?

File: src/models.jl (hunk 5 above)
Comment (bkamins) [reply src/models.jl:108]: I do not see it used except in line `H = H₀`

File: src/models.jl (hunk 1 above)
Comment (spagnuolocarmine) [reply src/models.jl:30]: We considered having at minimum to have self-loops hyperedges...

File: src/models.jl (hunk 2 above)
Comment (spagnuolocarmine) [reply src/models.jl]: Yes, you are right. COVID effect on myself. I changed

File: src/models.jl (hunk 3 above)
Comment (spagnuolocarmine) [reply src/models.jl]: same

File: src/models.jl (hunk 5 above)
Comment (spagnuolocarmine) [reply src/models.jl]: I changed removing the not needed code

File: src/models.jl (hunk 1 above)
Comment (spagnuolocarmine) [reply src/models.jl:30]: We can remove this constrain, and update the doc.
?

Comment (bkamins) [reply src/models.jl:30]: I am OK with what you proposed - just wanted to understand the constraint.

File: test/runtests.jl [hunk 9]
```diff
@@ -15,6 +15,29 @@ h1[4,3:4] .= 4.5
...
 h1[5,2] = 6.5
 
+@testset "SimpleHypergraphs building models    " begin
+
+    Hᵣ = random_model(5,5)
+    @test nhv(Hᵣ) == 5
+    @test nhe(Hᵣ) == 5
+    @test  all(length.(Hᵣ.v2he) .> 0)
+    @test  all(length.(Hᵣ.v2he) .<= 5)
+
```
Comment (pszufe) [inline test/runtests.jl:25]: Add tests for corner cases with `0` and `1` values of one parameter and normal values of the other.

File: src/models.jl [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+**The Algorithm**
+The algorithm starts with a random graph with 5 nodes and 5 edges. For this reason,
+the generated random graph has at least 5 nodes and 5 edges.
+It iteratively adds a node or a edge, according to a given parameter *p*,
+which defines the probability of creating a new node or a new hyperedge.
+
+More in detail, the connections with the new node/hyperedge are generated according to
+a preferential attachment policy defined by _p_.
+"""
+function random_preferential_model(nVertices::Int, p::Real)
```
[the remark this answers is not in the document]
Comment (spagnuolocarmine) [reply src/models.jl:92]: I think that could be better to follow the article in this version of the code., which uses a starting hypergraph of 5x5

File: src/models.jl [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+            for v in nextNodes(H,y)
+                push!(Y,v)
+            end
+        end
+            SimpleHypergraphs.add_hyperedge!(H, vertices=Y)
+    end
+    H
+end
+
+
```
[the remark this answers is not in the document]
Comment (spagnuolocarmine) [reply src/models.jl:114]: where?

File: src/models.jl [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+function nextNodes(h,size)
+    nodes = Dict{Int, Bool}()
+
+    ids = collect(1:nhv(h))
+    degrees = length.(h.v2he)
+
+    for s=1:size
+        psum = collect(1:length(ids))
```
[the remark this answers is not in the document]
Comment (spagnuolocarmine) [reply src/models.jl:122]: is possible to define as psum = Vector{Int}(undef, length(ids))
but must be initialized each loop, because it ids size changes during the loop.

File: src/models.jl [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+
+        psum[1] = degrees[ids[1]]
+        for j=2:length(ids)
+            psum[j] = psum[j-1] + degrees[ids[j]]
+        end
+
+        number = rand(1:psum[length(psum)])
+        bucket = -1
```
[the remark this answers is not in the document]
Comment (spagnuolocarmine) [reply src/models.jl:130]: done

File: src/models.jl (hunk 12 above)
[the remark this answers is not in the document]
Comment (pszufe) [reply src/models.jl]: Currently it is:
```
psum = collect(1:length(ids))
```
after executing this line the value of `psum` is `[1,2,3, ..., length(ids)]`
In the lines 136 and 138 you immedietely fill it in with different values. 

Actually since `ids` is reduced by one element throughout the loo, we can put `psum = Vector{Int}(undef, length(ids))` before the loop and run `resize!(psum,length(psum)-1)` at the end of each loop and then reuse `psum`. Is this something that you meat?

[the remark this answers is not in the document]
Comment (spagnuolocarmine) [reply src/models.jl]: the problem here is resize!(psum,length(psum)-1), you have to remove by one but according to the selected id as deleteat!(ids, index)

File: src/models.jl (hunk 10 above)
[the remark this answers is not in the document]
Comment (spagnuolocarmine) [reply src/models.jl:95]: we add this change.

File: src/models.jl (hunk 11 above)
[the remark this answers is not in the document]
Comment (spagnuolocarmine) [reply src/models.jl:117]: sure. we add this change.

**Comment by spagnuolocarmine** [comment]
We have followed all suggestions. Please merge with the master. If are you agreeing on it
